### PR TITLE
[SPARK-21144][SQL] Print a warning if the data schema and partition schema have the duplicate columns

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
The current master outputs unexpected results when the data schema and partition schema have the duplicate columns:
```
withTempPath { dir =>
  val basePath = dir.getCanonicalPath
  spark.range(0, 3).toDF("foo").write.parquet(new Path(basePath, "foo=1").toString)
  spark.range(0, 3).toDF("foo").write.parquet(new Path(basePath, "foo=a").toString)
  spark.read.parquet(basePath).show()
}

+---+
|foo|
+---+
|  1|
|  1|
|  a|
|  a|
|  1|
|  a|
+---+
```
This patch added code to print a warning when the duplication found.

## How was this patch tested?
Manually checked.

### CODE REVIEW

**Comment by maropu** [comment]
[This fix](https://github.com/apache/spark/pull/18375/files#diff-7a6cb188d2ae31eb3347b5629a679cecR196) breaks the existing test suites, so I'm looking for other approaches to fix this issue only...

**Comment by maropu** [comment]
@gatorsmile I remembered @liancheng said `we want to allow users to create partitioned tables that allow data schema to contain (part of) the partition columns, and there are test cases for this use case` before (https://github.com/apache/spark/pull/16030#issuecomment-263981907). But, I feel the query in the description seems to be error-prone, so how about just printing warning messages when detecting the duplication ([like here)](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Column.scala#L264)?

**Comment by gatorsmile** [comment]
uh, I see. Let us log it as a warning.

**Comment by maropu** [comment]
ok, I'll update.

**Comment by maropu** [comment]
@gatorsmile Manually checked:
```
scala> Seq((1, 2, 3)).toDF("a", "b", "c").write.save(s"$path/a=0")
scala> spark.read.load(path).show
17/06/23 21:04:44 WARN SchemaUtils: Found duplicate column(s) in the data schema and the partition schema: `a`. You might need to assign different column names.
+---+---+---+
|  a|  b|  c|
+---+---+---+
|  0|  2|  3|
+---+---+---+
```

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+   * Checks if input column names have duplicate identifiers. This throws an exception if
+   * the duplication exists.
+   *
+   * @param columnNames column names to check
+   * @param colType column type name, used in an exception message
+   * @param caseSensitiveAnalysis whether duplication checks should be case sensitive or not
+   */
+  def checkColumnNameDuplication(
+      columnNames: Seq[String], colType: String, caseSensitiveAnalysis: Boolean): Unit = {
+    val names = if (caseSensitiveAnalysis) {
```
Comment (brad-kaiser) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala]: This if might be easier to read on one line.
```scala
    val names = if (caseSensitiveAnalysis) columnNames else columnNames.map(_.toLowerCase)
```

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala]: Yeah. Let me first merge this.

@maropu , could you resolve this in your another related PR?

Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/util/SchemaUtils.scala]: ok

**Pull request closed without merging**